File: android/app/src/main/kotlin/com/bockmedia/console/local/HouseholdStore.kt
package com.bockmedia.console.local

import com.bockmedia.console.data.api.dto.HouseholdResponse
import com.bockmedia.console.data.repository.BockMediaRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Server-backed household member list — source of truth is GET /api/household. */
object HouseholdStore {
    private val _revision = MutableStateFlow(0)
    val revision: StateFlow<Int> = _revision.asStateFlow()

    @Volatile
    var cached: HouseholdResponse? = null
        private set

    fun members(): List<com.bockmedia.console.data.api.dto.HouseholdMember> =
        cached?.members.orEmpty()

    fun memberExists(id: String?): Boolean {
        val mid = id?.trim().orEmpty()
        if (mid.isEmpty()) return false
        return members().any { it.id == mid }
    }

    suspend fun refresh(repository: BockMediaRepository): HouseholdResponse {
        val fresh = runCatching { repository.household() }.getOrDefault(HouseholdResponse())
        apply(fresh)
        return fresh
    }

    fun apply(response: HouseholdResponse) {
        val prev = cached
        val changed = prev == null ||
            response.members != prev.members ||
            response.deviceOwners != prev.deviceOwners ||
            response.clientBindings != prev.clientBindings
        cached = response
        if (changed) _revision.value += 1
    }
}

File: ios/BockMedia/Data/Local/HouseholdStore.swift
import Foundation

/// Server-backed household member list — source of truth is `GET /api/household`.
@MainActor
enum HouseholdStore {
    static let changedNotification = Notification.Name("BockHouseholdChanged")

    private(set) static var cached: HouseholdResponse?
    private(set) static var revision = 0
    private(set) static var lastLoadError: String?

    static func members() -> [HouseholdMember] {
        cached?.members ?? []
    }

    static func memberExists(_ id: String?) -> Bool {
        guard let id, !id.isEmpty else { return false }
        return members().contains { $0.id == id }
    }

    @discardableResult
    static func refresh(repository: BockMediaRepository, force: Bool = false) async -> HouseholdResponse {
        if force {
            repository.invalidateAPI()
        }
        do {
            let fresh = try await repository.household()
            lastLoadError = nil
            apply(fresh)
            return fresh
        } catch {
            lastLoadError = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
            return cached ?? HouseholdResponse()
        }
    }

    static func apply(_ response: HouseholdResponse) {
        let changed = response.members != cached?.members
            || response.deviceOwners != cached?.deviceOwners
            || response.clientBindings != cached?.clientBindings
        cached = response
        guard changed else { return }
        revision += 1
        NotificationCenter.default.post(name: changedNotification, object: nil)
    }
}
